### ISSUE TITLE
How to make -storage.local.retention=24h longer

### ISSUE THREAD

**Issue by kubebc** [coreos/prometheus-operator#732]
Hi, Team,

1.  I've tried to update "retention" from 24h to 8760h below, but both failed. Any tips? Thanks!

$ kubectl edit statefulsets -n tectonic-system prometheus-k8s
```
apiVersion: apps/v1beta1
kind: StatefulSet
..
    spec:
      containers:
      - args:
        - -storage.local.retention=24h
```

Edited this file from console. 
```
kind: Prometheus
  name: k8s
  replicas: 1
  retention: 8760h
  ruleSelector:
```

2.  Are my storage config right?
```
  retention: 8760h
  storage:
    volumeClaimTemplate
      - metadata:
        name: prometheus-k8s-db
        annotations:
         volume.beta.kubernetes.io/storage-class: gold
      spec:
        accessModes:
        - ReadWriteOnce
        resources:
          requests:
            storage: 10Gi
        selector:
          matchLabels:
            app: prometheus-k8s
```

**Comment by brancz** [comment]
The Tectonic Prometheus instance intentionally does not offer persistent storage today, as we're about to migrate it to Prometheus 2.0. Regarding the retention, there is a `ConfigMap` called `tectonic-monitoring` in the `tectonic-system` namespace, within it you can configure selected things for the Tectonic Monitoring stack. We need this in order to be able to control the customization so we can provide any guarantee for automated upgrades to work. 

See the documentation for what is explicitly configurable [here](https://coreos.com/tectonic/docs/latest/tectonic-prometheus-operator/user-guides/configuring-tectonic-monitoring.html).

**Comment by kubebc** [comment]
Amazing! 
1. I've set "retention: 8760h" following your guide.
```
$ kubectl edit cm -n tectonic-system tectonic-monitoring
configmap "tectonic-monitoring" edited
$ kubectl edit statefulset -n tectonic-system prometheus-k8s
Edit cancelled, no changes made.
spec:
      containers:
      - args:
        - -storage.local.retention=8760h
```
2. prometheus-k8s-db cannot survive after pod restating, what's your recommendation to backup and recover prometheus-k8s-db? Thanks! 
```
      volumes:
      - emptyDir: {}
        name: prometheus-k8s-db

        volumeMounts:
        - mountPath: /var/prometheus/data
          name: prometheus-k8s-db
```
Fyi, I used to copy data files out from prometheus container as a backup, and then copy back in as recovery. But those data are unrecognizable at last.
```
$ kubectl exec -it prometheus-3067433533-4rz0q -- sh
/prometheus # tar zcvf data.tgz data
$ kubectl exec -it prometheus-3067433533-4rz0q -- sh
/prometheus # ls
data      data.tgz
/prometheus # du -sh data.tgz
43.7M    data.tgz
$ kubectl cp default/prometheus-3067433533-4rz0q:/prometheus/data.tgz data.tgz
tar: removing leading '/' from member names
$ ls
data.tgz
$ du -sh data.tgz
44M    data.tgz
```

**Comment by kubebc** [comment]
Can I add permanent storage the same way as "retention" at here?

$ kubectl edit cm -n tectonic-system tectonic-monitoring
```
apiVersion: v1
data:
  config.yaml: |
    prometheusOperator:
      baseImage: quay.io/coreos/prometheus-operator
      prometheusConfigReloaderBaseImage: quay.io/coreos/prometheus-config-reloader
      configReloaderBaseImage: quay.io/coreos/configmap-reload
    prometheusK8s:
      retention: 8760h
      storage:
        volumeClaimTemplate
          - metadata:
            name: prometheus-k8s-db
            annotations:
             volume.beta.kubernetes.io/storage-class: gold
          spec:
            accessModes:
            - ReadWriteOnce
            resources:
              requests:
                storage: 10Gi
            selector:
              matchLabels:
                app: prometheus-k8s
      baseImage: quay.io/prometheus/prometheus
    alertmanagerMain:
      baseImage: quay.io/prometheus/alertmanager
    ingress:
      baseAddress: base.address.com
kind: ConfigMap
metadata:
  creationTimestamp: 2017-08-24T08:35:14Z
  name: tectonic-monitoring
  namespace: tectonic-system
  resourceVersion: "35133205"
  selfLink: /api/v1/namespaces/tectonic-system/configmaps/tectonic-monitoring
  uid: 26b85bf4-88a7-11e7-a5f6-123c979a9fd4
```

**Comment by brancz** [comment]
Persistent storage for the prometheus-k8s instance is intentionally not configurable today, as we're about to migrate to Prometheus 2.0, which has a completely new storage format, and data migrations are close to impossible. Therefore we decided to only start allowing to configure persistent storage once we roll out Prometheus 2.0. I understand this is not ideal, but Prometheus 2.0 stable was released today, and we're already in the process of integrating it into Tectonic :slightly_smiling_face: .

**Comment by kubebc** [comment]
Thanks! brancz

1. We're now using tectonic_1.7.5. When can we expect to migrate to Prometheus 2.0 on our stack?

2. I just read release note and Prometheus 2.0 migration guide. Amazing! **We can upgrade to 2.0 on our current Tectonic stack, right? Would be no impact on Tectonic stack itself right?** 
https://coreos.com/blog/prometheus-2.0-released
https://github.com/prometheus/prometheus/releases/tag/v2.0.0
https://prometheus.io/docs/prometheus/2.0/migration/

3. This is the only one to access our historic monitoring data, right?
```
"To retain access to your historic monitoring data, run a non-scraping Prometheus instance running at least version 1.8.1 in parallel with your Prometheus 2.0 instance, and **have the new server read existing data from the old one via the remote read protocol**." 
```

**Comment by brancz** [comment]
It sounds like you've just rolled out Prometheus, if you don't have a lot of important historical data, I highly recommend just upgrading to 2.0 without data migration.

The Tectonic Monitoring Prometheus instance will automatically be upgraded for you in one of the next Tectonic releases, in facts it's what I'm working on as we speak.

**Comment by gmichels** [comment]
Sorry for sort of hijacking the thread but I was wondering how does one accomplish the same using just kube-prometheus. I don't see any config maps with this entry.

**Comment by brancz** [comment]
I'll close this here. If you have follow-up questions feel free to re-open or create a new issue.

Happy holidays! :tada:

**Comment by ScottBrenner** [comment]
As `retention` is not defined in `kube-prometheus` out-of-the-box, what does it default to?

**Comment by brancz** [comment]
Persistent storage is not available in the same way on all clusters, the kube prometheus stack is really just an example and is absolutely there for you to customize to your need.

Default retention is 24h iirc.
